src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs

```diff
@@ -995,6 +995,13 @@ private static MethodDesc TryResolveVirtualStaticMethodOnThisType(MetadataType c
                 if (methodImpl.Decl == interfaceMethodDefinition)
                 {
                     MethodDesc resolvedMethodImpl = methodImpl.Body;
+                    if (resolvedMethodImpl.OwningType != constrainedType.GetTypeDefinition())
```

Comment:
Note that the managed type system does a lot less validation than the CoreCLR type system in general - there are many other things that could be wrong with the type and prevent a load on CoreCLR that we'll not detect here. The way this is approached elsewhere is that we emit enough fixups to force this load on the CoreCLR side. I guess adding a check fixes some test, but it's not systematic.

Reply:
An example of the problem would be:

```csharp
Call<Foo>();

static void Call<T>() where T : IFoo => T.Something();

interface IFoo
{
    static virtual void Something() => Console.WriteLine("Hello");
}

struct Foo : IFoo
{
    extern void BreakThings();
}
```

Here CoreCLR will not allow `Foo` to load because it has an extern method with no body (just a random example of things CoreCLR looks at). If we generate code in crossgen2 that bypasses the need to load `Foo` (which I believe we'd do here, but I didn't actually check - I'm certain it can be tweaked to achieve that), we'd "fix" this even though it's not supposed to work.

Reply:
Hmm, can you please explain to me why the problem you described is specific to static virtual methods? I mean, let's say Crossgen2 allows JIT to inline a call to a static method - we're also not making sure on the Crossgen2 side that the callee sits in a loadable class that doesn't have any extern methods defined and whatnot. You may be right that it's problematic in certain corner cases and we should improve the type validation but I just don't see why this should be a SVM-specific problem; if a change in this logic turns out to be necessary, I tend to assume it should apply to all methods, not just static virtuals.

Reply:
I don't think it's specific to static virtual methods. It is a general problem. I think the throw/check added here is just enough to fix a failing test, but it doesn't even fix this problem for bad MethodImpls, let alone for the many other cases when CoreCLR would refuse to load the type. E.g. if there's a broken MethodImpl for some other method but it's _after_ we already found a match, we'll not notice and will not throw. CoreCLR will.

The preferred solution for this has always been to generate a fixup to force CoreCLR to load this type before we're allowed to run the code.

If a fixup is not possible, we're basically saying we'll need a CoreCLR-like type load emulator that will validate this whenever we touch a type. If that's the case, we should disable the test on ActiveIssue and budget for a CoreCLR-load emulator in .NET 9.

Adding partial checks in places that our tests happened to run into doesn't seem sustainable to me. E.g. we iterate MethodImpls in all of these places. Should each of these validate the MethodImpl is valid? And what about all the other scenarios where CoreCLR will refuse a load that we just happen not to have test coverage with static virtuals?

https://github.com/dotnet/runtime/blob/cdd7566751185d8330b1229b5b7aeb64605fbbf2/src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs#L287-L301

https://github.com/dotnet/runtime/blob/cdd7566751185d8330b1229b5b7aeb64605fbbf2/src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs#L492-L515

https://github.com/dotnet/runtime/blob/cdd7566751185d8330b1229b5b7aeb64605fbbf2/src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs#L794-L816

Reply:
Agreed. I've written the bare bones of a CoreCLR based type load emulator as part of supporting the skip type validation feature of R2R, but its incomplete, and does not cover all of the failure paths (It handles most of the checks that skip type validation skips). I see that we ALSO have a different validator that is used in NativeAOT/Crossgen2 in other cases in the jit interface. We really ought to unify the two and finish them up in the future.

Reply:
> I see that we ALSO have a different validator that is used in NativeAOT/Crossgen2 in other cases in the jit interface

That validator was originally written for NativeAOT purposes - we have a more relaxed approach to invalid inputs there and the validator basically just does enough validation to prevent a compiler crash later (type system exceptions are catchable/handlable only when hit while generating method code and fatal in most other places).

Crossgen2 needs to be more strict because "fixing" a problem initially, but resurfacing it again when the method is rejitted into a higher tier is worse than the NativeAOT behavior (that just doesn't throw the exception it was supposed to throw and leaves things at that).

Reply:
Both R2R and tiered JIT assume that the IL is valid. Invalid IL can have behavior differences between R2R and JIT or between Tier0 JIT and Tier1 JIT. The AOT compiler validator does not need to be perfect. Its main purpose is to catch missing dependencies that are very common in the code out there. We have tacked some invalid metadata handling to it too, but that is mostly just to make some of our own tests for invalid patterns happy.

Reply:
If we view it as "just a thing to make a test pass", I'm fine with adding a check here.

However, adding a new ExceptionStringId requires more work than just a new enum member. NativeAOT compiler will potentially catch this exception at compile time and rethrow it at runtime, generating an exception message using CoreLib's localization.

Also the exception type and exception message should match CoreCLR's exception (that's the pattern we've been following).